### PR TITLE
(v6) toJSON special cases

### PR DESCRIPTION
**Pull request by steffenagger**
## What, How & Why?
Added special cases for `Realm.Object.toJSON`:
- ignores all functions & internals
- converts `ArrayBuffer`s to base64 strings (currently these outputs as `{}`).
- *Added:* fix for https://github.com/realm/realm-js/issues/3331

## ☑️ ToDos

* [x] 📝 Changelog entry
* [ ] 🚦 Tests

### CODE REVIEW

File: lib/extensions.js [hunk 1]
```diff
@@ -162,10 +162,23 @@ module.exports = function(realmConstructor, context) {
             Object.keys(this)
                 .concat(Object.keys(Object.getPrototypeOf(this)))
                 .forEach(key => {
-                    const value = this[key];
-                    result[key] = value instanceof realmConstructor.Object || value instanceof realmConstructor.Collection
-                        ? value.toJSON(key, cache)
-                        : value;
+                    const org = this[key];
```
Comment (kraenhansen) [inline lib/extensions.js:165]: I personally prefer not to abbreviate local variables like this - I'll let you decide if you want this suggestion.

```suggestion
                    const original = this[key];
```

File: lib/extensions.js [hunk 2]
```diff
@@ -162,10 +162,23 @@ module.exports = function(realmConstructor, context) {
...
+
+                    if (org instanceof realmConstructor.Object || org instanceof realmConstructor.Collection) {
+                        // recursively trigger `toJSON` for Realm instances with the same cache.
+                        val = org.toJSON(key, cache);
+                    } else if (org instanceof ArrayBuffer) {
+                        // convert "data" to base64 strings.
+                        val = Buffer.from(org).toString("base64");
+                    }
+
+                    result[key] = val !== undefined ? val : org;
```
Comment (kraenhansen) [inline lib/extensions.js:181]: I think this change could make us a bit more robust to the global being swapped (as some Jest environments turns out to do).

```suggestion
                    result[key] = typeof val !== "undefined" ? val : org;
```

**Review by kraenhansen — CHANGES_REQUESTED** [review]

File: lib/extensions.js (hunk 1 above)
Comment (steffenagger) [reply lib/extensions.js:165]: Agreed. After creating the PR I gave it a workover.

**Review by kraenhansen — APPROVED** [review]

**Review by kneth — APPROVED** [review]
An entry in `CHANGELOG.md` would be nice 😃

**Review by kneth — APPROVED** [review]

**Review by blagoev — APPROVED** [review]

**Pull request merged**